### DISCUSSION TITLE
Environment variables not working with JS project

### DISCUSSION THREAD

**Discussion by breadadams** [community/community#25930] [Actions]
<p>I’m not sure if this is really the best place to post this, but going to do so anyway in case it helps anyone else facing the same issue.</p>
<p>So we have a JS project, and a github workflow that runs a build script, creates a zip, and uploads it to the Chrome web store. All parts of this have been working fine until we introduced environment variables.</p>
<p>When we run build locally process.env.X is replaced correctly with the contents of our .env file, however in the files generated from the action it displays as e.env.X (and the application doesn’t work).</p>
<p>We are using dotenv via the dotenv-webpack plugin. Maybe we need to set the path to the env file with this, if so, what would said path be? I looked into it but didn’t find any information.</p>
<p>Here’s the action file:</p>
<pre><code>name: Upload

on:
  release:
    types: [published]

jobs:
  upload:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
      - run: yarn
      - run: yarn build
        env:
          WEB_APP_URL: https://someurl.com/
          WELCOME_URL: https://someurl.com/
      - uses: JasonEtco/upload-to-release@v0.1.1
        with:
          args: chrome-extension.zip application/zip
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      - uses: Klemensas/chrome-extension-upload-action@v1.3
        with:
          refresh-token: ${{ secrets.REFRESH_TOKEN }}
          client-id: ${{ secrets.CLIENT_ID }}
          client-secret: ${{ secrets.CLIENT_SECRET }}
          file-name: chrome-extension.zip
          app-id: ${{ secrets.APP_ID }}
</code></pre>
<p>Note that the <em>variables</em> for the upload-to-release &amp; chrome-extension-upload-action steps work fine, it’s just those on the yarn build command.</p>
<p>If I can share anything else that would be of use to debug this let me know! Thanks</p>

**Comment by weide-zhou** [comment]
<p>Command ‘yarn build’ is not an existing command by default, ‘build’ is a command which can be specified in your package.json file on the scripts property.</p>
<p>Typically we can use env in workflow like below:</p>
<pre><code>run: yarn build --all -u $BUILD_TOKEN
      env:
        BUILD_TOKEN: ${{ secrets.BUILD_TOKEN }}
</code></pre>

**Comment by breadadams** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>, thanks for the response, and yes I’m aware that build isn’t a default command/script.</p>
<p>I’m not too sure about the “–all -u” flags in your example, is that something specific to <em>your</em> build? (I can’t seem to find those flags in the yarn docs).</p>
<p>Anyway, so I’ve tried your suggestion with the following workflow yaml, but still the zip file that’s generated is missing the env variables like I described in my initial post.</p>
<pre><code>name: Env Test

on: push

jobs:
  upload:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
      - run: yarn
      - run: yarn build $WEB_APP_URL $WELCOME_URL
        env:
          WEB_APP_URL: https://example.com/
          WELCOME_URL: https://example.com/
      - uses: actions/upload-artifact@v1
        with:
          name: example
          path: example.zip
</code></pre>
<p>Am I missing something?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/breadadams">@breadadams</a> ,</p>
<p>Thanks for your reply! Please ignore ‘–all -u’ flags. After further investigation, ‘yarn build’ should be able to accept env values. Sample as below, In Node.js you can access it via  <strong>process.env.API_KEY</strong>. More details <a href="https://stackoverflow.com/questions/59686823/how-can-i-use-github-secrets-in-js-files" rel="nofollow noopener">here</a>.</p>
<pre><code>- name: Run Tests (JEST)
        run: yarn test --ci --silent --testPathIgnorePatterns=experimental
        env:
          CI: true
          API_KEY: ${{ secrets.API_KEY }}
</code></pre>
<p>Please check your js file about the ENV. If the error still persists, please kinldy share a repo/sample repo for further investigation.</p>
<p>Thanks.</p>

**Comment by breadadams** [comment]
<p>Hey <a class="mention" href="/weide-zhou">@weide-zhou</a>, excuse the delay, not been able to prepare a sample repo.</p>
<p>I’ve stripped down the code &amp; dependencies to absolute basics from our project and created this repo: <a href="https://github.com/breadadams/gh-action-env-trial" rel="nofollow noopener">https://github.com/breadadams/gh-action-env-trial</a></p>
<p>The instructions in the README should be clear, let me know if you have any questions about running it.</p>

**Comment by breadadams** [comment, marked as answer]
<p>The issue was with our configuration for the dotenv-webpack plugin, given that the environment  variables aren’t coming from a <em>real</em> <code>.env</code> file we needed to enable the <strong>systemvars</strong> setting, as seen here: <a href="https://github.com/breadadams/gh-action-env-trial/commit/5d673f3a55e54b14344995d6f5d5f039dc6f56d4" rel="nofollow noopener">https://github.com/breadadams/gh-action-env-trial/commit/5d673f3a55e54b14344995d6f5d5f039dc6f56d4</a></p>
